天気予報API機能完成

tamaron opened the pull request.

tamaron committed:
入力したURL先が存在するならtextで表示するように

tamaron committed:
LitJsonの追加　「天気」というキーワードで現在の天気を返すようにした

tamaron committed:
OpenWeatherMapに鞍替えした

tamaron committed:
Merge pull request #9 from tamaron/feature/AddWebAPI

Feature/add web api

hanachiru commented:
まずGameManagerについて，`GameManager`と`TextGroupView`に相互依存の関係があってあまり良くない構造だと思う．

基本的に`GameManager`は他のクラスから使われる側のクラスであって，他のクラスのメソッドを呼んだりとかは神クラスって言われるようなクラスになりがちかも (そもそも主語が大きいのが要因かもしれない)

https://github.com/tamaron/QQAgent/blob/7b15de0c6dd4e0d41273479c2fdc7e57f3f59bc0/Assets/Scripts/GameManager.cs#L8-L40

今のGameManagerの責務大きく分けて書いてみると
1. Gameの進行の状態を保持
1. `TextGroupView`のPopUp呼び出し
2. 状態の変更
3. inputGroup, textGruup関連のSetActive等

あたりだとと思う．

できればGameManagerはGameの進行に関する状態を保持・変更するだけのクラスにしておいた方がシンプルになる気がする．

例えばPopUpに関しては，`TextGroupView.cs`で下みたいにする感じ．
> GameManager.Instance.State.Where(s => s == GameState.WaitingInput).Subscribe(_ => PopUp());

後inputGroup, textGruupのSetActive関連に関してはGameManagerとは別のクラスを作って，そこで`Stateを購読 -> SetActiveとか諸々`とかをした方がより責務を分散できるかも．

hanachiru commented:
次にMV(R)Pに関してだけれど，UIと相性は抜群だし僕は結構好みです．

TextGroupView.cdに関して，カプセル化の観点から以下のフィールドを`public`にするのは良くないかも．せっかくViewクラスを定義してるなら，見た目・ユーザー入力関連の責務はこのクラスに完結すべきかな．
https://github.com/tamaron/QQAgent/blob/7b15de0c6dd4e0d41273479c2fdc7e57f3f59bc0/Assets/Scripts/View/TextGroupView.cs#L17-L18

今の状態だとその責務をPresenterに押し付けちゃってる感じ．

https://github.com/tamaron/QQAgent/blob/7b15de0c6dd4e0d41273479c2fdc7e57f3f59bc0/Assets/Scripts/Presenter/TextGroupPresenter.cs#L20-L29

例えば以下の感じに変えたりも良いかも．
> TextGroupView.cs
> [SerializeField] private TextMeshProUGUI resultText;
> public IObservable\<string\> OnInputTextAsObservable() => resultText.onEndEdit.AsObservable();

こうすることで，他のクラスにテキストの変更とか色々なことができちゃったのをテキストを入力したときに発火するストリームの購読だけをさせることができます．
`_view.resultText.text = ----`も`SetText`みたいにViewクラスにメソッドを定義してそれを呼び出す方がいいかな．

後TextGroupView.csには
1. ユーザーからの入力
2. 生成した応答文の出力

の２種類の責務があって，今の規模だと同じクラスでも良いけれどViewを分けるのも将来的にはアリだと思う．

将来的に入力はInputFieldだけじゃなくて、音声入力も，出力はテキストだけじゃなくて音声出力とかもとかになると肥大化していっちゃう可能性はあるかな．

hanachiru commented:
次にModelに関して．処理の流れとしてはよく出来てると思います．

ただやっぱり責務が詰め込みすぎてるっていうのはあるかもしれません．
1. 入力文に対してどのタイプで対応するか
2. APIにリクエストを送る
3. 応答文生成

どこまでクラスを分けるかは悩ましいですが，少なくとも以下の2つはクラスを分けた方が良いと思います．(このままだとクラスが肥大化しちゃう)
- MessageTypeに応じた処理の変更
- MessageTypeに応じた応答文生成

例えば天気予報の応答文を生成するクラスをつくるなら
> Weather〇〇.cs
> 1. APIを叩く
> 2. 応答文生成

みたいな感じですかね．

tamaron commented:
ありがとうございます
後でゆっくり見させていただきます！

hanachiru commented:
補足でやや発展的なことですが，クラス・メソッドを適切に分けるとテストがしやすくなるというメリットもあります．

MV(R)Pを適応した理由にも挙げられますが，例えば用例応答生成が正しく動作しているかだけを調べたいとします．(逆にViewやPresenterにバグがあっても構わない)

MV(R)Pパターンを用いていれば，ロジックを分けられるので用例応答のテストも容易になります．
> ReplyTest.cs
> 1. var replyGenerater = new TextGroupModel()
> 2. var reply = replayGenerator.GetResultAsync("今日の天気を教えて")
> 3. replyがちゃんと天気関連のものになっているか調べる

もっと続けていくと，APIとのやりとりがちゃんとしているかのテスト，正しいMessageTypeに分けられているかのテスト...みたいに分けられるかもしれません．

MonoBehaviorに依存してるとテストがしづらいのも事実ですが，それ用のシーンを用意してテストするのも全然アリだと思います．

またViewのテストも同様ですが，テスト用のPresenterクラス・Modelクラスを用意して，動作を確認する(用例応答の場合なら，入力に対してそのまま返すクラスを定義するとか)なんてこともできます．

まあテストを書くのが面倒いってのもありますが．

後GameManager(ゲームの進行に関する状態の管理クラス)もModelといってもいいのかもしれませんね．
ただ利便性的にSingletonにしたいのはすごい分かるし，お好みで．

tamaron committed:
GameManeger, View, Presenterの修正

tamaron commented:
レビューありがとうございます！
指摘いただいた箇所のいくつかを修正しました．

- GameManagerクラスの分割

`GameManager`を3つのクラスに分割しました．（画面の初期化をする`CameraInitializer`クラス， ゲームの状態を保持する`GameStateModel`クラス，ゲームの状態が変わったときに実行する処理を登録する`GameManager`クラス）
また，PopUp処理の購読はView側にまかせるようにしました．
`GameStateModel`は`ReactiveProperty<GameState>`型のフィールドだけを持つクラスであり`MonoBehaviour`を継承する必要がないため継承しないようにしました．（ただし参照しやすいようにSingletonのままにしています）



- Viewの分割

`TextGroupView`をPopUpの処理や購読を管理する`TextGroupPopup`クラス，ユーザからの入力を受け取る`TextGroupInput`クラス，Agentからの出力を表示する`TextGroupOutput`クラスの3つに分割しました．

- Viewの責務をPresenterに押し付けないように



PresenterがViewのフィールドを`_view.resultText.text = ○○○○;`のように直接変更していたため
```
    public string ResultText
    {
        get { return outputText.text; }
        set { outputText.text = value; }
    }
```
のようにプロパティを介して変更するようにしました．

また，View側でユーザの入力が終わったときに発行される`IObservable<string>`型のメソッドを定義して，PresenterがViewにあまり踏み込みすぎないようにしました．

後ほどModelの方も分割していこうと思います．

tamaron committed:
model分割 名前空間の定義(一部)

もともとTextViewModelが担っていた，入力文の識別，応答分生成の責務をそれぞれMessageTypeCategorizer，ResultTextGeneratorに分散させた

tamaron committed:
名前空間の定義

tamaron committed:
応答文生成待機時間の調整

tamaron commented on Assets/Scripts/Model/TextGroupModel.cs:
```diff
@@ -10,63 +10,18 @@
-    {
-        if (Regex.IsMatch(text, "天気")) return MessageType.Weather;
-        return MessageType.None;
-    }
```
`TextGroupModel`がPresenterから入力を受け取ったときの挙動を
- `MessageTypeCategorizer`に入力文を渡す．`MessageTypeCategorizer.Generator`に適切なタイプの`ResultTextGenerator`が代入される(実際には`ResultTextGenerator`を継承した`NoneGenerator`や`WeatherGenerator`が入る)
- `ResultTextGenerator.Generate()`を実行する．`ResultTextGenerator.Result`に応答文が入る
という流れにすることでModelの分割をしました．
